File: agent/internal/protocol/protocol.go
// Package protocol mirrors the Control Plane's wire contract in
// protocols/v2 — the envelope and message payload shapes exchanged over
// /ws/agent, and the closed set of structured command types. There is no
// shell/free-form command anywhere in this protocol.
package protocol

import (
	"encoding/json"
	"time"
)

// Version is the protocol_version this Agent speaks — see
// protocols/v2/envelope.schema.json.
const Version = "1.0"

// Envelope wraps every message exchanged over the WebSocket connection.
type Envelope struct {
	ProtocolVersion string          `json:"protocol_version"`
	MessageID       string          `json:"message_id"`
	Type            string          `json:"type"`
	Timestamp       time.Time       `json:"timestamp"`
	Payload         json.RawMessage `json:"payload"`
}

// Message types.
const (
	TypeAgentHello        = "agent.hello"
	TypeAgentHeartbeat    = "agent.heartbeat"
	TypeAgentCommandEvent = "agent.command_event"
	TypeControlCommand    = "control.command"
)

// HelloPayload is sent once, immediately after connecting — the capability
// report. See protocols/v2/agent-hello.schema.json.
type HelloPayload struct {
	AgentVersion string   `json:"agent_version"`
	OS           string   `json:"os"`
	Arch         string   `json:"arch"`
	Adapters     []string `json:"adapters,omitempty"`
}

// InstanceStatus is one entry in a heartbeat's per-instance status list.
type InstanceStatus struct {
	InstanceID string `json:"instance_id"`
	Status     string `json:"status"`
	Port       int    `json:"port,omitempty"`
}

// HeartbeatPayload is the periodic metrics snapshot. See
// protocols/v2/agent-heartbeat.schema.json.
type HeartbeatPayload struct {
	CPUPercent    float64          `json:"cpu_percent"`
	MemoryPercent float64          `json:"memory_percent"`
	DiskPercent   float64          `json:"disk_percent"`
	Instances     []InstanceStatus `json:"instances"`
}

// CommandEnvelope is the payload of a control.command message. See
// protocols/v2/command-envelope.schema.json.
type CommandEnvelope struct {
	CommandID      string          `json:"command_id"`
	IdempotencyKey string          `json:"idempotency_key"`
	Type           string          `json:"type"`
	Payload        json.RawMessage `json:"payload"`
	CreatedAt      time.Time       `json:"created_at"`
	ExpiresAt      *time.Time      `json:"expires_at"`
	CorrelationID  *string         `json:"correlation_id"`
}

// Command event statuses.
const (
	StatusAcknowledged = "acknowledged"
	StatusRunning      = "running"
	StatusSucceeded    = "succeeded"
	StatusFailed       = "failed"
	StatusTimedOut     = "timed_out"
)

// CommandEvent reports command progress/result. See
// protocols/v2/command-event.schema.json.
type CommandEvent struct {
	CommandID  string          `json:"command_id"`
	Status     string          `json:"status"`
	Result     json.RawMessage `json:"result,omitempty"`
	Error      *string         `json:"error,omitempty"`
	OccurredAt time.Time       `json:"occurred_at"`
}

// The ten structured command types the Agent will ever be asked to run.
const (
	CommandInspectHost     = "inspect_host"
	CommandValidateApp     = "validate_app"
	CommandDeployRelease   = "deploy_release"
	CommandStartInstance   = "start_instance"
	CommandStopInstance    = "stop_instance"
	CommandRestartInstance = "restart_instance"
	CommandInspectInstance = "inspect_instance"
	CommandCollectLogs     = "collect_logs"
	CommandCollectMetrics  = "collect_metrics"
	CommandUpdateProxy     = "update_proxy"
)

// KnownCommandTypes is the closed set the Control Plane may ever send.
// Anything not in this set is rejected as "unknown" rather than merely
// "unimplemented" — see internal/dispatcher.
var KnownCommandTypes = map[string]bool{
	CommandInspectHost:     true,
	CommandValidateApp:     true,
	CommandDeployRelease:   true,
	CommandStartInstance:   true,
	CommandStopInstance:    true,
	CommandRestartInstance: true,
	CommandInspectInstance: true,
	CommandCollectLogs:     true,
	CommandCollectMetrics:  true,
	CommandUpdateProxy:     true,
}

File: services/control-plane/app/ws/protocol.py
"""Python-side mirror of the JSON message contract in protocols/v1.

The JSON Schemas under protocols/v1 remain the source of truth for the wire
format; these are the runtime validation/construction helpers used by the
Control Plane.
"""

import uuid
from datetime import UTC, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

PROTOCOL_VERSION = "1.0"


def make_envelope(message_type: str, payload: dict[str, Any]) -> dict[str, Any]:
    return {
        "protocol_version": PROTOCOL_VERSION,
        "message_id": str(uuid.uuid4()),
        "type": message_type,
        "timestamp": datetime.now(UTC).isoformat(),
        "payload": payload,
    }


class AgentHelloPayload(BaseModel):
    """Sent once, immediately after the WebSocket connects — the capability
    report. See protocols/v1/agent-hello.schema.json.
    """

    agent_version: str
    os: Literal["windows", "linux"]
    arch: Literal["amd64", "arm64"]
    adapters: list[str] = Field(default_factory=list)


class AgentHeartbeatPayload(BaseModel):
    cpu_percent: float = Field(ge=0, le=100)
    memory_percent: float = Field(ge=0, le=100)
    disk_percent: float = Field(ge=0, le=100)
    instances: list[dict[str, Any]] = Field(default_factory=list)


class AgentCommandEventPayload(BaseModel):
    command_id: uuid.UUID
    status: Literal["acknowledged", "running", "succeeded", "failed", "timed_out"]
    result: dict[str, Any] | None = None
    error: str | None = None
    occurred_at: datetime
